pkg/cloudprovider/providers/vsphere/nodemanager.go

```diff
@@ -20,33 +20,168 @@ import (
 	"k8s.io/api/core/v1"
 	k8stypes "k8s.io/apimachinery/pkg/types"
 	"k8s.io/kubernetes/pkg/cloudprovider/providers/vsphere/vclib"
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"strings"
+	"errors"
+	"sync"
 )
 
 // Stores info about the kubernetes node
 type NodeInfo struct {
+	dataCenter *vclib.Datacenter
 	vm       *vclib.VirtualMachine
 	vcServer string
 }
 
 type NodeManager struct {
+	// TODO: replace map with concurrent map when k8s supports go v1.9
+
 	// Maps the VC server to VSphereInstance
 	vsphereInstanceMap map[string]*VSphereInstance
 	// Maps node name to node info.
 	nodeInfoMap map[string]*NodeInfo
+	// Maps node name to node structure
+	registeredNodes map[string]*v1.Node
+
+
+	// Mutexes
+	registeredNodesLock sync.RWMutex
+	nodeInfoLock sync.RWMutex
+}
+
+var (
+	ErrNoNodeInfo = errors.New("no node info")
+	ErrNoVSphereInstance = errors.New("no vsphere instance found")
+)
+
+func (nm *NodeManager) DiscoverNode(node *v1.Node) error {
+	for vc, vsi := range nm.vsphereInstanceMap {
```

Comment:
We are looping all VCs serially but we can discover nodes across VCs in parallel. Any reason why we don't want to do that? I don't know if there executor service (Java) equivalent in go, I'll let you know if we can use thread pool concept in go if I come across a library that does that.

Reply:
Good Catch! Will figure this out